[plugin:vite:import-analysis] Failed to resolve import "http-proxy-agent" from "node_modules/.vite/deps/ccxt.js?v=9c98d43a". Does the file exist?

### Operating System

_No response_

### Programming Languages

JavaScript

### CCXT Version

4.2.60

### Description

Vue3.3+vite4.4.9 run error

[plugin:vite:import-analysis] Failed to resolve import "http-proxy-agent" from "node_modules/.vite/deps/ccxt.js?v=9c98d43a". Does the file exist?

### Code

```
  
import ccxt from "ccxt";

const ccxtInstance = ccxt
const exchange = ccxtInstance.binance
```

> I managed to get the latest version to run on Vite (with Solid.js in this case).
> 
> 1. `npm i https-proxy-agent`
> 2. For vite launch dev mode `optimizeDeps` seems to be required.
> 3. For a build to succeed it was necessary to add `build.rollupOptions`
> 
> I hope this saves you some time and nerves ✌️
> 
> Here is my vite config file:
> 
> ```js
> import { defineConfig } from "vite";
> import solidPlugin from "vite-plugin-solid";
> 
> import { nodePolyfills } from "vite-plugin-node-polyfills";
> export default defineConfig({
>   optimizeDeps: {
>     include: ["http-proxy-agent"],
>   },
>   plugins: [
>     solidPlugin(),
>     nodePolyfills({
>       protocolImports: true,
>     }),
>   ],
>   server: {
>     port: 4000,
>   },
>   build: {
>     target: "esnext",
>     rollupOptions: {
>       external: ["http-proxy-agent"],
>     },
>   },
> });
> ```

This does solves the 'http-proxy-agent' issue but then now I get this instead, I guess it's something wrong with the polyfills:

```
✘ [ERROR] No matching export in "node_modules/.pnpm/node-stdlib-browser@1.2.0/node_modules/node-stdlib-browser/esm/mock/empty.js" for import "isIP"

    node_modules/.pnpm/ccxt@4.3.60/node_modules/ccxt/js/src/static_dependencies/node-fetch/utils/referrer.js:7:9:
      7 │ import { isIP } from 'node:net';
        ╵          ~~~~
```

> > I managed to get the latest version to run on Vite (with Solid.js in this case).
> > 
> > 1. `npm i https-proxy-agent`
> > 2. For vite launch dev mode `optimizeDeps` seems to be required.
> > 3. For a build to succeed it was necessary to add `build.rollupOptions`
> > 
> > I hope this saves you some time and nerves ✌️
> > Here is my vite config file:
> > ```js
> > import { defineConfig } from "vite";
> > import solidPlugin from "vite-plugin-solid";
> > 
> > import { nodePolyfills } from "vite-plugin-node-polyfills";
> > export default defineConfig({
> >   optimizeDeps: {
> >     include: ["http-proxy-agent"],
> >   },
> >   plugins: [
> >     solidPlugin(),
> >     nodePolyfills({
> >       protocolImports: true,
> >     }),
> >   ],
> >   server: {
> >     port: 4000,
> >   },
> >   build: {
> >     target: "esnext",
> >     rollupOptions: {
> >       external: ["http-proxy-agent"],
> >     },
> >   },
> > });
> > ```
> 
> This does solves the 'http-proxy-agent' issue but then now I get this instead, I guess it's something wrong with the polyfills:
> 
> ```
> ✘ [ERROR] No matching export in "node_modules/.pnpm/node-stdlib-browser@1.2.0/node_modules/node-stdlib-browser/esm/mock/empty.js" for import "isIP"
> 
>     node_modules/.pnpm/ccxt@4.3.60/node_modules/ccxt/js/src/static_dependencies/node-fetch/utils/referrer.js:7:9:
>       7 │ import { isIP } from 'node:net';
>         ╵          ~~~~
> ```

I created a patch over `node-stdlib-browser`, in their index.js they export `empty.js` as net even though they have mock implementation for the net package. I just replaced the path and it's running.

> > > I managed to get the latest version to run on Vite (with Solid.js in this case).
> > > 
> > > 1. `npm i https-proxy-agent`
> > > 2. For vite launch dev mode `optimizeDeps` seems to be required.
> > > 3. For a build to succeed it was necessary to add `build.rollupOptions`
> > > 
> > > I hope this saves you some time and nerves ✌️
> > > Here is my vite config file:
> > > ```js
> > > import { defineConfig } from "vite";
> > > import solidPlugin from "vite-plugin-solid";
> > > 
> > > import { nodePolyfills } from "vite-plugin-node-polyfills";
> > > export default defineConfig({
> > >   optimizeDeps: {
> > >     include: ["http-proxy-agent"],
> > >   },
> > >   plugins: [
> > >     solidPlugin(),
> > >     nodePolyfills({
> > >       protocolImports: true,
> > >     }),
> > >   ],
> > >   server: {
> > >     port: 4000,
> > >   },
> > >   build: {
> > >     target: "esnext",
> > >     rollupOptions: {
> > >       external: ["http-proxy-agent"],
> > >     },
> > >   },
> > > });
> > > ```
> > 
> > 
> > This does solves the 'http-proxy-agent' issue but then now I get this instead, I guess it's something wrong with the polyfills:
> > ```
> > ✘ [ERROR] No matching export in "node_modules/.pnpm/node-stdlib-browser@1.2.0/node_modules/node-stdlib-browser/esm/mock/empty.js" for import "isIP"
> > 
> >     node_modules/.pnpm/ccxt@4.3.60/node_modules/ccxt/js/src/static_dependencies/node-fetch/utils/referrer.js:7:9:
> >       7 │ import { isIP } from 'node:net';
> >         ╵          ~~~~
> > ```
> 
> I created a patch over `node-stdlib-browser`, in their index.js they export `empty.js` as net even though they have mock implementation for the net package. I just replaced the path and it's running.

@WRadoslaw  hi, how to fix this?

--- a/ts/src/base/Exchange.ts
+++ b/ts/src/base/Exchange.ts
@@ -166,7 +166,7 @@ const {
 export type { Market, Trade, Fee, Ticker, OHLCV, OHLCVC, Order, OrderBook, Balance, Balances, Dictionary, Transaction, Currency, MinMax, IndexType, NullableIndexType, Int, Bool, OrderType, OrderSide, Position, LedgerEntry, BorrowInterest, OpenInterest, LeverageTier, TransferEntry, CrossBorrowRate, FundingRateHistory, Liquidation, FundingHistory, OrderRequest, MarginMode, Tickers, Greeks, Option, OptionChain, Str, Num, MarketInterface, CurrencyInterface, BalanceAccount, MarginModes, MarketType, Leverage, Leverages, LastPrice, LastPrices, Account, Strings, Conversion, DepositAddress, LongShortRatio, ADL } from './types.js';
 // ----------------------------------------------------------------------------
 //
-const dynamicImport = async (moduleName: string) => await import (/* webpackIgnore: true */ moduleName);
+const dynamicImport = async (moduleName: string) => await import (/* webpackIgnore: true */ /* @vite-ignore */ moduleName); // both bundler hints needed so optional proxy-agent modules are not statically resolved, see https://github.com/ccxt/ccxt/issues/21555
 //
 // ----------------------------------------------------------------------------
 //

